### PR TITLE
Add automated test suite to the library.

### PR DESCRIPTION
**Pull request by jamestalmage**
This module is hugely popular, it really should have some proper tests. This patch adds them.

The old tests are moved to the `test/manual` folder.

Run the new ones with `npm test`

The tests all run locally using a real express server. I don't want the default test suite to query remote servers. We could create an `e2e` suite that does hit real servers, just for a sanity check, but I'm not sure that is even necessary.

~~It currently provides 100% coverage, but that is a bit misleading as it never makes an `https` call (all testing is currently via `http`). I intend to do that, but it's not trivial. It's going to involve creating self certificates for the client, server, and a certificate authority.~~

UPDATE: now includes some tests for https

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -102,10 +102,12 @@ for (var protocol in protocols) {
           var searchname = parsedUrl.search || '';
           var hostname = parsedUrl.hostname;
           var pathname = parsedUrl.pathname;
+          var port = parsedUrl.port;
 
           var redirectOptions = typeof options === 'string' ? {} : options;
           redirectOptions.reqUrl = redirectUrl;
           redirectOptions.hostname = hostname;
+          redirectOptions.port = port;
```
Comment (jamestalmage) [inline index.js]: Proper testing is already paying dividends. I've discovered we were not following redirect urls correctly.

**Comment by olalonde** [comment]
Wow, thanks a lot @jamestalmage, it's really great you are doing this as I'm pretty overloaded right now and don't even use this module myself anymore. I added you as a contributor so feel free to merge to master. Just ping me when you need me to `npm publish`!

**Comment by olalonde** [comment]
Oh, also, feel free to rename my old "tests" or rename it to "examples".

**Comment by jamestalmage** [comment]
Thank you! I really appreciate your trust. There are definitely a few places where I think I can make some improvements, but I wanted a good test suite before making any significant changes.

**Pull request merged**